_posts/2020-02-11-spring-boot-openapi.md

```diff
@@ -0,0 +1,394 @@
+
+If you take a quick look under `target/generated-sources/openapi/src/main/java/io/reflectoring/model`, you will find the User model defined in `openapi.yml`!
+
+```java
+@javax.annotation.Generated(value = "org.openapitools.codegen.languages.SpringCodegen", date = "2020-02-07T19:50:29.578+01:00[Europe/Berlin]")
+
+public class User   {
+  @JsonProperty("id")
+  private Long id;
+
+  @JsonProperty("username")
+  private String username;
+
+  @JsonProperty("firstName")
+  private String firstName;
+
+  @JsonProperty("lastName")
+  private String lastName;
+
+  @JsonProperty("email")
+  private String email;
+
+  @JsonProperty("password")
+  private String password;
+
+  @JsonProperty("phone")
+  private String phone;
+
+  @JsonProperty("userStatus")
+  private Integer userStatus;
+}
+// omit getters and setters
+```
+
+Now let's implement the `UserApi`, which comes from our generated specification.
+
+```java
+@RestController
+public class UserController implements UserApi {
+}
+```
+
+If we test our API without having implemented a method of `UserApi`, we'll get a 501 response (Not Implemented). Why?
+
+```shell script
+curl -I http://localhost:8080/user/Petros
+HTTP/1.1 501
+Content-Length: 0
+```
+
+Because in UserApi says so:
+
+```java
+default ResponseEntity<User> getUserByName(String username) {
+    getRequest().ifPresent(request -> {
+        for (MediaType mediaType: MediaType.parseMediaTypes(request.getHeader("Accept"))) {
+            if (mediaType.isCompatibleWith(MediaType.valueOf("application/json"))) {
+                String exampleString = "{ \"firstName\" : \"firstName\", \"lastName\" : \"lastName\", \"password\" : \"password\", \"userStatus\" : 6, \"phone\" : \"phone\", \"id\" : 0, \"email\" : \"email\", \"username\" : \"username\" }";
+                ApiUtil.setExampleResponse(request, "application/json", exampleString);
+                break;
+            }
+        }
+    });
+    return new ResponseEntity<>(HttpStatus.NOT_IMPLEMENTED);
+
+}
+```
+
+Let's now just implement our actual GET method:
+
+```java
+@Override
+public ResponseEntity<User> getUserByName(String username) {
+    User user = new User();
+
+    user.setId(123L);
+    user.setFirstName("Petros");
+    user.setLastName("S");
+    user.setUsername("Petros");
+    user.setEmail("petors.stergioulas94@gmail.com");
+    user.setPassword("secret");
+    user.setPhone("+123 4567890");
+    user.setUserStatus(0);
+
+    return ResponseEntity.ok(user);
+}
+```
+Running now: `curl http://localhost:8080/user/Petros`
+
+You will receive a valid response!
+
+```json
+{
+  "id": 123,
+  "username": "Petros",
+  "firstName": "Petros",
+  "lastName": "S",
+  "email": "petors.stergioulas94@gmail.com",
+  "password": "secret",
+  "phone": "+123 4567890",
+  "userStatus": 0
+}
+```
+
+In my opinion, generating the OpenAPI Specification with Maven plugin instead of Swagger Editor is a better choice. 
```

Comment:
Can you elaborate a bit on why this is a good solution? Why do we have more control? What does "safely track any changes" mean? Which classes are we supposed to touch manually (the controller). What will happen if we generate code again (we have to manually update the controller, but not the other classes)?